revocation/crl/cache/file.go

```diff
@@ -0,0 +1,248 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"archive/tar"
+	"context"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/notaryproject/notation-core-go/revocation/internal/file"
+)
+
+const (
+	// pathBaseCRL is the file name of the base CRL
+	pathBaseCRL = "base.crl"
+
+	// pathMetadata is the file name of the metadata
+	pathMetadata = "metadata.json"
+
+	// tempFileName is the prefix of the temporary file
+	tempFileName = "notation-*"
+)
+
+// FileCache stores in a tarball format, which contains two files: base.crl and
```

Comment:
Why do we need tarball? Can we store everything in single file? Name that file based on sha256 hash of URI and all other information is inside the file.

Reply:
There is a critical reason that we need a tarball. The CRL and delta CRL are a bundle, and updating the bundle should be an atomic operation, or we may end up in an inconsistent state during concurrent operations.

For example, when the CRL has reached `Next Update`, both the base CRL and delta CRL will be refreshed. If Process A is updating these two files in the cache while Process B is reading the cache, it is possible that the base CRL is not updated but the delta CRL is, or vice versa. This inconsistency makes the CRL and delta CRL out of sync, requiring the program to handle more complicated cases. By using a tarball, we can make the operation atomic and avoid these issues.

Reply:
Also, to store the download URL and Cached At time, we need a metadata file which also needs to be updated with CRL bundle atomically.

Reply:
Removed `FileCache`

Reply:
> This inconsistency makes the CRL and delta CRL out of sync, requiring the program to handle more complicated cases. By using a tarball, we can make the operation atomic and avoid these issues.

In case of out of sync we should just refresh the case with next sane value.

Reply:
What about `metadata` for each CRL? It should be updated in a bundle with the CRL atomically.

Reply:
Removed `Metadata`. We can use your proposed cache solution.